nomt/src/io/page_pool.rs

```diff
@@ -158,44 +189,63 @@ impl PagePool {
         self.inner.freelist.write().push(page);
     }
 
-    fn grow(&self, freelist: &mut Vec<Page>) -> Page {
-        assert!(freelist.is_empty());
+    fn grow(&self, freelist_guard: &mut RwLockWriteGuard<Vec<Page>>) -> Page {
+        assert!(freelist_guard.is_empty());
+
+        // First step is to allocate a new region.
         let region_ptr = unsafe {
             libc::mmap(
                 std::ptr::null_mut(),
-                REGION_SIZE,
+                REGION_BYTE_SIZE,
                 libc::PROT_READ | libc::PROT_WRITE,
                 libc::MAP_PRIVATE | libc::MAP_ANONYMOUS,
                 /* fd */ -1,
                 /* offset */ 0,
             )
         };
+        assert!(!region_ptr.is_null());
```

Comment:
can you elaborate a bit?

Reply:
this is a bug actually, below this line we should check for the MAP_FAILED which is actually returned. This is more of a hint that whatever we store in the `regions` field is not null.

Reply:
ah, and in turn, we want the region ptrs, the allocated ones anyway, be non-null.

Why? Because that compromises safety (in the sense of UB):

1. you don't want to construct a slice from a ptr which is null.
2. you don't want to compute offset from null.

Now, when I was looking for a proof, I wasn't able to find the exact doc line stating that you cannot create offsets against null. However, I just remember that from the experience. Here is the proof

https://godbolt.org/z/GnvsfbdPr

It might be possible that the compiler does it because it expects that each pointer points to a distinct "object" (for some definition of an object) and no other object could be reached through that pointer and null pointer is special, and in such a case where the compiler doesn't know that the ptr is statically always null then it won't pull out any UB.

But I think I spent way too much of my time even to write this message, so I would rather leave that check in place kek

Reply:
Thank you!